ibis/file/csv.py

```diff
@@ -0,0 +1,111 @@
+import pandas as pd
+import ibis.expr.operations as ops
+from ibis.file.client import FileClient
+from ibis.pandas.core import pre_execute, execute  # noqa
+from ibis.pandas.client import pandas_dtypes_to_ibis_schema
+from ibis.pandas.execution.selection import physical_tables
+
+
+def connect(path):
+    """Create a CSVClient for use with Ibis
+
+    Parameters
+    ----------
+    path: str or pathlib.Path
+
+    Returns
+    -------
+    CSVClient
+    """
+
+    return CSVClient(path)
+
+
+class CSVTable(ops.DatabaseTable):
+    pass
+
+
+class CSVClient(FileClient):
+    extension = 'csv'
+
+    def insert(self, path, t, index=False, **kwargs):
+        path = self.root / path
+        data = execute(t)
+        data.to_csv(str(path), index=index, **kwargs)
+
+    def table(self, name, path=None):
+        if name not in self.list_tables(path):
+            raise AttributeError(name)
+
+        if path is None:
+            path = self.root
+
+        # get the schema
+        f = path / "{}.{}".format(name, self.extension)
+        df = pd.read_csv(str(f), header=0, nrows=10)
+        schema = pandas_dtypes_to_ibis_schema(df, {})
+
+        t = CSVTable(name, schema, self).to_expr()
+        self.dictionary[name] = f
+        return t
+
+    def list_tables(self, path=None):
+        # tables are files in a dir
+        if path is None:
+            path = self.root
+
+        tables = []
+        if path.is_dir():
+            for d in path.iterdir():
+                if d.is_file():
+                    if str(d).endswith(self.extension):
+                        tables.append(d.stem)
+        elif path.is_file():
+            if str(path).endswith(self.extension):
+                tables.append(path.stem)
+        return tables
+
+    def list_databases(self, path=None):
+        # databases are dir
+        if path is None:
+            path = self.root
+
+        tables = []
+        if path.is_dir():
+            for d in path.iterdir():
+                if d.is_dir():
+                    tables.append(d.name)
+        return tables
+
+
+@pre_execute.register(CSVTable, CSVClient)
+def csv_pre_execute_table(op, client, scope=None, **kwargs):
```

Comment:
You'll want to make sure that `op` isn't already in scope before doing this otherwise you'll end up reading the file again if this table is traversed in the execute loop. Something like:

```python
if op not in scope:
    # read csv etc
else:  # we're already in scope
    return scope
```

Reply:
We can think about maybe making this the default behavior.

Reply:
But in a follow up issue/PR

Reply:
done, yes I agree we can do this at a higher level